### ISSUE TITLE
Untracked files in git repository after running "flutter create ."

### ISSUE THREAD

**Issue by jsroest** [flutter/flutter#133323]
### Steps to reproduce:
Run in an empty folder:  
1. flutter create .
1. git init
1. git add .
1. git commit -m "initial commit"
1. git clean -xdn
1. git ls-files . --ignored --exclude-standard --others

### Problem
The output of the last two commands show a bunch of untracked files and folders created by the `flutter create .` command. This is not desirable, because now there are differences between a fresh checkout and a freshly created project.

### Fix
1. Fix the output of all flutter create template types (app, module, package, plugin, plugin_ffi and skeleton) to only generate files that are committed to git or adjust the gitignore to include them.
2. Guard this by a github action that fails if the output of `git clean -xdn` or `git ls-files . --ignored --exclude-standard --others` lists untracked files.

### Related issues
#9460

### Logs

<details>
<summary>flutter create .</summary>

```console  
Creating project ....
Resolving dependencies... (1.1s)
Got dependencies.
Wrote 108 files.

All done!
You can find general documentation for Flutter at: https://docs.flutter.dev/
Detailed API documentation is available at: https://api.flutter.dev/
If you prefer video documentation, consider: https://www.youtube.com/c/flutterdev

In order to run your application, type:

  $ cd .
  $ flutter run

Your application code is in ./lib/main.dart.
```

</details>

<details>
<summary>git init</summary>

```console
Initialized empty Git repository in /Users/johannesroest/repos/flutter_git_test/.git/
```

</details>

<details>
<summary>git add .</summary>

```console
```

</details>

<details>
<summary>git commit -m "initial commit"</summary>

```console
[master (root-commit) 6189338] initial commit
 101 files changed, 3379 insertions(+)
 create mode 100644 .gitignore
 create mode 100644 .metadata
 create mode 100644 README.md
 create mode 100644 analysis_options.yaml
 create mode 100644 android/.gitignore
 create mode 100644 android/app/build.gradle
 create mode 100644 android/app/src/debug/AndroidManifest.xml
 create mode 100644 android/app/src/main/AndroidManifest.xml
 create mode 100644 android/app/src/main/kotlin/com/example/flutter_git_test/MainActivity.kt
 create mode 100644 android/app/src/main/res/drawable-v21/launch_background.xml
 create mode 100644 android/app/src/main/res/drawable/launch_background.xml
 create mode 100644 android/app/src/main/res/mipmap-hdpi/ic_launcher.png
 create mode 100644 android/app/src/main/res/mipmap-mdpi/ic_launcher.png
 create mode 100644 android/app/src/main/res/mipmap-xhdpi/ic_launcher.png
 create mode 100644 android/app/src/main/res/mipmap-xxhdpi/ic_launcher.png
 create mode 100644 android/app/src/main/res/mipmap-xxxhdpi/ic_launcher.png
 create mode 100644 android/app/src/main/res/values-night/styles.xml
 create mode 100644 android/app/src/main/res/values/styles.xml
 create mode 100644 android/app/src/profile/AndroidManifest.xml
 create mode 100644 android/build.gradle
 create mode 100644 android/gradle.properties
 create mode 100644 android/gradle/wrapper/gradle-wrapper.properties
 create mode 100644 android/settings.gradle
 create mode 100644 ios/.gitignore
 create mode 100644 ios/Flutter/AppFrameworkInfo.plist
 create mode 100644 ios/Flutter/Debug.xcconfig
 create mode 100644 ios/Flutter/Release.xcconfig
 create mode 100644 ios/Runner.xcodeproj/project.pbxproj
 create mode 100644 ios/Runner.xcodeproj/project.xcworkspace/contents.xcworkspacedata
 create mode 100644 ios/Runner.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist
 create mode 100644 ios/Runner.xcodeproj/project.xcworkspace/xcshareddata/WorkspaceSettings.xcsettings
 create mode 100644 ios/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme
 create mode 100644 ios/Runner.xcworkspace/contents.xcworkspacedata
 create mode 100644 ios/Runner.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist
 create mode 100644 ios/Runner.xcworkspace/xcshareddata/WorkspaceSettings.xcsettings
 create mode 100644 ios/Runner/AppDelegate.swift
 create mode 100644 ios/Runner/Assets.xcassets/AppIcon.appiconset/Contents.json
 create mode 100644 ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-1024x1024@1x.png
 create mode 100644 ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-20x20@1x.png
 create mode 100644 ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-20x20@2x.png
 create mode 100644 ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-20x20@3x.png
 create mode 100644 ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-29x29@1x.png
 create mode 100644 ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-29x29@2x.png
 create mode 100644 ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-29x29@3x.png
 create mode 100644 ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-40x40@1x.png
 create mode 100644 ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-40x40@2x.png
 create mode 100644 ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-40x40@3x.png
 create mode 100644 ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-60x60@2x.png
 create mode 100644 ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-60x60@3x.png
 create mode 100644 ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-76x76@1x.png
 create mode 100644 ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-76x76@2x.png
 create mode 100644 ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-83.5x83.5@2x.png
 create mode 100644 ios/Runner/Assets.xcassets/LaunchImage.imageset/Contents.json
 create mode 100644 ios/Runner/Assets.xcassets/LaunchImage.imageset/LaunchImage.png
 create mode 100644 ios/Runner/Assets.xcassets/LaunchImage.imageset/LaunchImage@2x.png
 create mode 100644 ios/Runner/Assets.xcassets/LaunchImage.imageset/LaunchImage@3x.png
 create mode 100644 ios/Runner/Assets.xcassets/LaunchImage.imageset/README.md
 create mode 100644 ios/Runner/Base.lproj/LaunchScreen.storyboard
 create mode 100644 ios/Runner/Base.lproj/Main.storyboard
 create mode 100644 ios/Runner/Info.plist
 create mode 100644 ios/Runner/Runner-Bridging-Header.h
 create mode 100644 ios/RunnerTests/RunnerTests.swift
 create mode 100644 lib/main.dart
 create mode 100644 macos/.gitignore
 create mode 100644 macos/Flutter/Flutter-Debug.xcconfig
 create mode 100644 macos/Flutter/Flutter-Release.xcconfig
 create mode 100644 macos/Flutter/GeneratedPluginRegistrant.swift
 create mode 100644 macos/Runner.xcodeproj/project.pbxproj
 create mode 100644 macos/Runner.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist
 create mode 100644 macos/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme
 create mode 100644 macos/Runner.xcworkspace/contents.xcworkspacedata
 create mode 100644 macos/Runner.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist
 create mode 100644 macos/Runner/AppDelegate.swift
 create mode 100644 macos/Runner/Assets.xcassets/AppIcon.appiconset/Contents.json
 create mode 100644 macos/Runner/Assets.xcassets/AppIcon.appiconset/app_icon_1024.png
 create mode 100644 macos/Runner/Assets.xcassets/AppIcon.appiconset/app_icon_128.png
 create mode 100644 macos/Runner/Assets.xcassets/AppIcon.appiconset/app_icon_16.png
 create mode 100644 macos/Runner/Assets.xcassets/AppIcon.appiconset/app_icon_256.png
 create mode 100644 macos/Runner/Assets.xcassets/AppIcon.appiconset/app_icon_32.png
 create mode 100644 macos/Runner/Assets.xcassets/AppIcon.appiconset/app_icon_512.png
 create mode 100644 macos/Runner/Assets.xcassets/AppIcon.appiconset/app_icon_64.png
 create mode 100644 macos/Runner/Base.lproj/MainMenu.xib
 create mode 100644 macos/Runner/Configs/AppInfo.xcconfig
 create mode 100644 macos/Runner/Configs/Debug.xcconfig
 create mode 100644 macos/Runner/Configs/Release.xcconfig
 create mode 100644 macos/Runner/Configs/Warnings.xcconfig
 create mode 100644 macos/Runner/DebugProfile.entitlements
 create mode 100644 macos/Runner/Info.plist
 create mode 100644 macos/Runner/MainFlutterWindow.swift
 create mode 100644 macos/Runner/Release.entitlements
 create mode 100644 macos/RunnerTests/RunnerTests.swift
 create mode 100644 pubspec.lock
 create mode 100644 pubspec.yaml
 create mode 100644 test/widget_test.dart
 create mode 100644 web/favicon.png
 create mode 100644 web/icons/Icon-192.png
 create mode 100644 web/icons/Icon-512.png
 create mode 100644 web/icons/Icon-maskable-192.png
 create mode 100644 web/icons/Icon-maskable-512.png
 create mode 100644 web/index.html
 create mode 100644 web/manifest.json
```

</details>

<details>
<summary>git clean -xdn</summary>

```console
Would remove .dart_tool/
Would remove .idea/
Would remove android/app/src/main/java/
Would remove android/flutter_git_test_android.iml
Would remove android/gradle/wrapper/gradle-wrapper.jar
Would remove android/gradlew
Would remove android/gradlew.bat
Would remove android/local.properties
Would remove flutter_git_test.iml
Would remove ios/Flutter/Generated.xcconfig
Would remove ios/Flutter/flutter_export_environment.sh
Would remove ios/Runner/GeneratedPluginRegistrant.h
Would remove ios/Runner/GeneratedPluginRegistrant.m
Would remove macos/Flutter/ephemeral/
```

</details>

<details>
<summary>git ls-files . --ignored --exclude-standard --others</summary>

```console
.dart_tool/dartpad/web_plugin_registrant.dart
.dart_tool/package_config.json
.dart_tool/package_config_subset
.dart_tool/version
.idea/libraries/Dart_SDK.xml
.idea/libraries/KotlinJavaRuntime.xml
.idea/modules.xml
.idea/runConfigurations/main_dart.xml
.idea/workspace.xml
android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java
android/flutter_git_test_android.iml
android/gradle/wrapper/gradle-wrapper.jar
android/gradlew
android/gradlew.bat
android/local.properties
flutter_git_test.iml
ios/Flutter/Generated.xcconfig
ios/Flutter/flutter_export_environment.sh
ios/Runner/GeneratedPluginRegistrant.h
ios/Runner/GeneratedPluginRegistrant.m
macos/Flutter/ephemeral/Flutter-Generated.xcconfig
macos/Flutter/ephemeral/flutter_export_environment.sh
```

</details>

<details>
<summary>Flutter doctor</summary>

```console
[✓] Flutter (Channel stable, 3.13.1, on macOS 13.5 22G74 darwin-x64, locale en-NL)
    • Flutter version 3.13.1 on channel stable at /Users/johannesroest/fvm/versions/3.13.1
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision e1e47221e8 (2 days ago), 2023-08-22 21:43:18 -0700
    • Engine revision b20183e040
    • Dart version 3.1.0
    • DevTools version 2.25.0

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/johannesroest/Library/Android/sdk
    • Platform android-34, build-tools 32.1.0-rc1
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] VS Code (version 1.79.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (3 available)
    • MC3300x (mobile) • 22152523027000 • android-arm64  • Android 11 (API 30)
    • macOS (desktop)  • macos          • darwin-x64     • macOS 13.5 22G74 darwin-x64
    • Chrome (web)     • chrome         • web-javascript • Google Chrome 116.0.5845.110

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```
</details>

**Comment by dam-ease** [comment]
> This is not desirable, because now there are differences between a fresh checkout and a freshly created project

Hi @jsroest. Thanks for filing this issue. Can you explain more about the differences? I can't seem to see a difference between these.

**Comment by jsroest** [comment]
Some files are not checked in, so they will not be there if you do a fresh checkout of the repository.

For example, if you work on a project with a coworker, your coworker will miss these files if he checks out the repository.

- .dart_tool/dartpad/web_plugin_registrant.dart
- .dart_tool/package_config.json
- .dart_tool/package_config_subset
- .dart_tool/version
- .idea/libraries/Dart_SDK.xml
- .idea/libraries/KotlinJavaRuntime.xml
- .idea/modules.xml
- .idea/runConfigurations/main_dart.xml
- .idea/workspace.xml
- android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java
- android/flutter_git_test_android.iml
- android/gradle/wrapper/gradle-wrapper.jar
- android/gradlew
- android/gradlew.bat
- android/local.properties
- flutter_git_test.iml
- ios/Flutter/Generated.xcconfig
- ios/Flutter/flutter_export_environment.sh
- ios/Runner/GeneratedPluginRegistrant.h
- ios/Runner/GeneratedPluginRegistrant.m
- macos/Flutter/ephemeral/Flutter-Generated.xcconfig
- macos/Flutter/ephemeral/flutter_export_environment.sh

**Comment by dam-ease** [comment]
Thanks for your response. These seem to be files specified in `.gitignore`, can you check them out?

**Comment by BLKKKBVSIK** [comment]
> Thanks for your response. These seem to be files specified in `.gitignore`, can you check them out?

I think @jsroest would like the `.gitignore` generated by default to include those files.

- android/gradlew
- android/gradlew.bat

I agree with those files. The gradle documentation says to commit the wrapper.
But, all the `.dart_tool/*` & `.idea/*` shouldn't be committed IMO, since it's mostly caching and user-specific configuration.

**Comment by stuartmorgan** [comment]
Most of those ignored files are created by the second step in the output log (`Resolving dependencies...`), which prepares the project for things like using an IDE, or using native tooling. This is intentional, because it makes the fresh project more immediately usable by the developer. If you want to skip that step, you can pass `--no-pub`.

> now there are differences between a fresh checkout and a freshly created project.

That is correct; `flutter` creates various untracked files on the fly as a normal part of its operation.

> For example, if you work on a project with a coworker, your coworker will miss these files if he checks out the repository.

Your co-worker will get their own versions of these files as soon as they interact with the repository (e.g., `flutter get` or `flutter run`).

Closing as these files are intentionally ignored, and also intentionally created post-template-instantiation as a convenience.

**Comment by jsroest** [comment]
Thanks @stuartmorgan for the clear explanation. It makes total sense to me know. I was under the assumption that "flutter create" would only create files, and that "flutter pub get" would only download dependencies mentioned in the pubspec.yaml. Thanks for the insight!

**Comment by jsroest** [comment]
> > Thanks for your response. These seem to be files specified in `.gitignore`, can you check them out?
> 
> I think @jsroest would like the `.gitignore` generated by default to include those files.
> 
> * android/gradlew
> * android/gradlew.bat
> 
> I agree with those files. The gradle documentation says to commit the wrapper. But, all the `.dart_tool/*` & `.idea/*` shouldn't be committed IMO, since it's mostly caching and user-specific configuration.

Hi @BLKKKBVSIK,

I made some tooling to create a new flutter app containing our business template app, but with the flutter version of your choice. Just to be sure that all support files are created with the right flutter version. This works fine for a few years now, but for some time I noticed that some changes I made to ‘.idea’ and ‘iml’ files got automatically reverted. In the process of debugging that I noticed that flutter made some untracked files. I was not aware that these files are also created by a ‘flutter pub get’ command and assumed that to be a bug. I always thought that Android Studio or IntelliJ created the ‘.idea’ and ‘iml’ files on the fly. But I was wrong and learned something new 😎.
